internal/controllers/topology/cluster/structuredmerge/dryrun.go

```diff
@@ -0,0 +1,282 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package structuredmerge
+
+import (
+	"encoding/json"
+	"fmt"
+	"reflect"
+	"strings"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"sigs.k8s.io/cluster-api/internal/contract"
+)
+
+// getTopologyManagedFields returns metadata.managedFields entry tracking
+// server side apply operations for the topology controller.
+func getTopologyManagedFields(original client.Object) map[string]interface{} {
+	r := map[string]interface{}{}
+	for _, m := range original.GetManagedFields() {
+		if m.Operation == metav1.ManagedFieldsOperationApply &&
```

Comment:
We may also have to accept `metav1.ManagedFieldsOperationUpdate` for the objects which get created by our topology controller.